// terriajs/dist/Traits/SearchProviders/LocationSearchProviderTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import primitiveArrayTrait from "../Decorators/primitiveArrayTrait";
import primitiveTrait from "../Decorators/primitiveTrait";
import mixTraits from "../mixTraits";
import ModelTraits from "../ModelTraits";
import SearchProviderTraits from "./SearchProviderTraits";
export default class LocationSearchProviderTraits extends mixTraits(SearchProviderTraits) {
    url = "";
    recommendedListLength = 5;
    flightDurationSeconds = 1.5;
    isOpen = true;
    attributions = [];
}
__decorate([
    primitiveTrait({
        type: "string",
        name: "URL",
        description: "The URL of search provider."
    })
], LocationSearchProviderTraits.prototype, "url", void 0);
__decorate([
    primitiveTrait({
        type: "number",
        name: "recommendedListLength",
        description: "Maximum amount of entries in the suggestion list."
    })
], LocationSearchProviderTraits.prototype, "recommendedListLength", void 0);
__decorate([
    primitiveTrait({
        type: "number",
        name: "URL",
        description: "Time to move to the result location.",
        isNullable: true
    })
], LocationSearchProviderTraits.prototype, "flightDurationSeconds", void 0);
__decorate([
    primitiveTrait({
        type: "boolean",
        name: "Is open",
        description: "True if the search results of this search provider are visible; otherwise, false.",
        isNullable: true
    })
], LocationSearchProviderTraits.prototype, "isOpen", void 0);
__decorate([
    primitiveArrayTrait({
        type: "string",
        name: "Attribution",
        description: "The attribution text for this search provider.",
        isNullable: true
    })
], LocationSearchProviderTraits.prototype, "attributions", void 0);
export class SearchProviderMapCenterTraits extends ModelTraits {
    mapCenter = true;
}
__decorate([
    primitiveTrait({
        type: "boolean",
        name: "Map center",
        description: "Whether the current location of the map center is supplied with search request"
    })
], SearchProviderMapCenterTraits.prototype, "mapCenter", void 0);
