// terriajs/dist/Models/Catalog/CatalogItems/TileMapServiceCatalogItem.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { computed, makeObservable, observable, runInAction } from "mobx";
import TileMapServiceImageryProvider from "terriajs-cesium/Source/Scene/TileMapServiceImageryProvider";
import CatalogMemberMixin from "../../../ModelMixins/CatalogMemberMixin";
import MappableMixin from "../../../ModelMixins/MappableMixin";
import TileMapServiceCatalogItemTraits from "../../../Traits/TraitsClasses/TileMapServiceCatalogItemTraits";
import CreateModel from "../../Definition/CreateModel";
import proxyCatalogItemUrl from "../proxyCatalogItemUrl";
/**
 * A catalog item that loads a Tile Map Service imagery tileset created using
 * MapTiler or gdal2tiles.
 */
export default class TileMapServiceCatalogItem extends MappableMixin(CatalogMemberMixin(CreateModel(TileMapServiceCatalogItemTraits))) {
    static type = "tms";
    constructor(...args) {
        super(...args);
        makeObservable(this);
    }
    imageryProvider;
    get type() {
        return TileMapServiceCatalogItem.type;
    }
    async forceLoadMapItems() {
        let imageryProvider;
        if (this.url) {
            imageryProvider = await TileMapServiceImageryProvider.fromUrl(proxyCatalogItemUrl(this, this.url), {
                minimumLevel: this.minimumLevel,
                maximumLevel: this.maximumLevel,
                tileWidth: this.tileWidth,
                tileHeight: this.tileHeight,
                credit: this.attribution
            });
        }
        runInAction(() => {
            this.imageryProvider = imageryProvider;
        });
    }
    get mapItems() {
        return this.imageryProvider
            ? [
                {
                    imageryProvider: this.imageryProvider,
                    show: this.show,
                    alpha: this.opacity,
                    clippingRectangle: this.clipToRectangle
                        ? this.cesiumRectangle
                        : undefined
                }
            ]
            : [];
    }
}
__decorate([
    observable
], TileMapServiceCatalogItem.prototype, "imageryProvider", void 0);
__decorate([
    computed
], TileMapServiceCatalogItem.prototype, "mapItems", null);
