// terriajs/dist/Models/NoViewer.js
import Rectangle from "terriajs-cesium/Source/Core/Rectangle";
import CameraView from "./CameraView";
import GlobeOrMap from "./GlobeOrMap";
class NoViewer extends GlobeOrMap {
    static type = "none";
    type = NoViewer.type;
    terria;
    canShowSplitter = false;
    _currentView = new CameraView(Rectangle.MAX_VALUE);
    constructor(terriaViewer, _container) {
        super();
        this.terria = terriaViewer.terria;
    }
    destroy() { }
    doZoomTo(v, _t) {
        if (v instanceof CameraView) {
            this._currentView = v;
        }
        else if (v instanceof Rectangle) {
            this._currentView = new CameraView(v);
        }
        return Promise.resolve();
    }
    setInitialView(view) {
        this._currentView = view;
    }
    notifyRepaintRequired() { }
    pickFromLocation(_latLngHeight, _providerCoords, _existingFeatures) { }
    getCurrentCameraView() {
        return this._currentView;
    }
    getContainer() {
        return undefined;
    }
    pauseMapInteraction() { }
    resumeMapInteraction() { }
    _addVectorTileHighlight(_imageryProvider, _rectangle) {
        return () => { };
    }
}
export default NoViewer;
