// terriajs/dist/Traits/TraitsClasses/TileMapServiceCatalogItemTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { traitClass } from "../Trait";
import mixTraits from "../mixTraits";
import CatalogMemberTraits from "./CatalogMemberTraits";
import ImageryProviderTraits from "./ImageryProviderTraits";
import LayerOrderingTraits from "./LayerOrderingTraits";
import LegendOwnerTraits from "./LegendOwnerTraits";
import MappableTraits from "./MappableTraits";
import UrlTraits from "./UrlTraits";
let TileMapServiceCatalogItemTraits = class TileMapServiceCatalogItemTraits extends mixTraits(ImageryProviderTraits, LayerOrderingTraits, UrlTraits, MappableTraits, CatalogMemberTraits, LegendOwnerTraits) {
};
TileMapServiceCatalogItemTraits = __decorate([
    traitClass({
        description: `Creates one catalog item for a Tile Map Service (TMS) imagery tileset.`,
        example: {
            type: "tms",
            name: "Natural Earth II (TMS)",
            url: "https://storage.googleapis.com/terria-datasets-public/basemaps/natural-earth-tiles"
        }
    })
], TileMapServiceCatalogItemTraits);
export default TileMapServiceCatalogItemTraits;
