// terriajs/dist/Traits/TraitsClasses/WebMapTileServiceCatalogItemTraits.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import anyTrait from "../Decorators/anyTrait";
import objectArrayTrait from "../Decorators/objectArrayTrait";
import objectTrait from "../Decorators/objectTrait";
import primitiveTrait from "../Decorators/primitiveTrait";
import mixTraits from "../mixTraits";
import ModelTraits from "../ModelTraits";
import { traitClass } from "../Trait";
import CatalogMemberTraits from "./CatalogMemberTraits";
import GetCapabilitiesTraits from "./GetCapabilitiesTraits";
import ImageryProviderTraits from "./ImageryProviderTraits";
import LayerOrderingTraits from "./LayerOrderingTraits";
import LegendOwnerTraits from "./LegendOwnerTraits";
import LegendTraits from "./LegendTraits";
import MappableTraits from "./MappableTraits";
import UrlTraits from "./UrlTraits";
export class WebMapTileServiceAvailableStyleTraits extends ModelTraits {
    identifier;
    title;
    abstract;
    legend;
    isDefault = false;
}
__decorate([
    primitiveTrait({
        type: "string",
        name: "Style Identifier",
        description: "The identifier of the style."
    })
], WebMapTileServiceAvailableStyleTraits.prototype, "identifier", void 0);
__decorate([
    primitiveTrait({
        type: "string",
        name: "Title",
        description: "The title of the style."
    })
], WebMapTileServiceAvailableStyleTraits.prototype, "title", void 0);
__decorate([
    primitiveTrait({
        type: "string",
        name: "Abstract",
        description: "The abstract describing the style."
    })
], WebMapTileServiceAvailableStyleTraits.prototype, "abstract", void 0);
__decorate([
    objectTrait({
        type: LegendTraits,
        name: "Style Name",
        description: "The name of the style."
    })
], WebMapTileServiceAvailableStyleTraits.prototype, "legend", void 0);
__decorate([
    primitiveTrait({
        type: "string",
        name: "Is Default",
        description: "True if this Style is default; otherwise, false."
    })
], WebMapTileServiceAvailableStyleTraits.prototype, "isDefault", void 0);
export class WebMapTileServiceAvailableLayerStylesTraits extends ModelTraits {
    layerName;
    styles;
}
__decorate([
    primitiveTrait({
        type: "string",
        name: "Layer Name",
        description: "The name of the layer for which styles are available."
    })
], WebMapTileServiceAvailableLayerStylesTraits.prototype, "layerName", void 0);
__decorate([
    objectArrayTrait({
        type: WebMapTileServiceAvailableStyleTraits,
        name: "Styles",
        description: "The styles available for this layer.",
        idProperty: "identifier"
    })
], WebMapTileServiceAvailableLayerStylesTraits.prototype, "styles", void 0);
let WebMapTileServiceCatalogItemTraits = class WebMapTileServiceCatalogItemTraits extends mixTraits(LayerOrderingTraits, GetCapabilitiesTraits, ImageryProviderTraits, UrlTraits, MappableTraits, CatalogMemberTraits, LegendOwnerTraits) {
    isGeoServer = false;
    layer;
    style;
    availableStyles;
    parameters;
    requestEncoding = "RESTful";
};
__decorate([
    primitiveTrait({
        type: "string",
        name: "Is GeoServer",
        description: "True if this WMS is a GeoServer; otherwise, false."
    })
], WebMapTileServiceCatalogItemTraits.prototype, "isGeoServer", void 0);
__decorate([
    primitiveTrait({
        type: "string",
        name: "Layer",
        description: "The layer to display."
    })
], WebMapTileServiceCatalogItemTraits.prototype, "layer", void 0);
__decorate([
    primitiveTrait({
        type: "string",
        name: "Style",
        description: "The style to use with `Layer`."
    })
], WebMapTileServiceCatalogItemTraits.prototype, "style", void 0);
__decorate([
    objectArrayTrait({
        type: WebMapTileServiceAvailableLayerStylesTraits,
        name: "Available Styles",
        description: "The available styles.",
        idProperty: "layerName"
    })
], WebMapTileServiceCatalogItemTraits.prototype, "availableStyles", void 0);
__decorate([
    anyTrait({
        name: "Parameters",
        description: "Additional parameters to pass to the MapServer when requesting images."
    })
], WebMapTileServiceCatalogItemTraits.prototype, "parameters", void 0);
__decorate([
    primitiveTrait({
        type: "string",
        name: "Encoding",
        description: "The encoding of the tile images. We will try to load the tile images with this encoding, if not available we will fallback to KVP. Supported values are KVP and Restful"
    })
], WebMapTileServiceCatalogItemTraits.prototype, "requestEncoding", void 0);
WebMapTileServiceCatalogItemTraits = __decorate([
    traitClass({
        description: `Creates a single item in the catalog from a url that points to a wmts service.`,
        example: {
            type: "wmts",
            id: "a unique id for wmts example",
            name: "wmts example",
            url: "https://services.arcgisonline.com/arcgis/rest/services/Reference/World_Boundaries_and_Places/MapServer/WMTS/1.0.0/WMTSCapabilities.xml",
            layer: "Reference_World_Boundaries_and_Places",
            opacity: 1
        }
    })
], WebMapTileServiceCatalogItemTraits);
export default WebMapTileServiceCatalogItemTraits;
